### ISSUE TITLE
CREATE_OF / CONVERT_OF signed inconsistency

### ISSUE THREAD

**Issue by Timmmm** [rems-project/sail#198]
Looking at the CREATE_OF and CONVERT_OF macros in `sail.h`/`sail.c`, there are a few weird things about them:

1. There's a fair bit of redundancy; E.g. what's the difference between `create_mach_int_of_sail_int()` and `convert_mach_int_of_sail_int()`? In theory nothing.
2. Except... one does an unsigned conversion and one does signed!!
3. It feels like quite an overuse of macros. Makes it kind of hard to follow. Why not just write out the functions?

Also it would be nice if there were some comments saying what the various macros are supposed to do. It's not clear what the difference between `CREATE_OF` and `CONVERT_OF` or `COPY` and `RECREATE` are supposed to be. It would help if the grammar was more normal too, e.g. `create_foo_*from*_bar`, `convert_bar_*to*_foo`, etc.

**Comment by Alasdair** [comment]
I'm not sure the compiler currently generates CREATE_OF so that might be a bug.

Right now each macro corresponds to an instruction in the Sail intermediate representation here: https://github.com/rems-project/sail/blob/sail2/language/jib.ott, so it's roughly something like below:

| IR | stack type | non-stack type |
| --- | --- | --- |
| `x : T` | `x = CREATE(T)` | `T x; CREATE(T)(&x)` |
| `x = y` (same types) | `x = y` | `COPY(&x, y)` |
| `x = y` (different types) | `x = CONVERT_OF(type_of(x), type_of(y))(y)` | `CONVERT_OF(type_of(x), type_of(y))(&x, y)` |
| `clear x T` | no-op | `KILL(T)(&x)` |

Currently `x : T = y` is compiled as `x : T; x = y`, which I think is where `CREATE_OF` used to be used. `RECREATE` is an optimisation, if we know that another variable is no longer used we can re-use it's memory rather than freeing and re-allocating the space using `KILL` followed by `CREATE`. The IR -> IR analysis that does this doesn't really want to care about how exactly the types are represented in C, so for some types `RECREATE`  and `CREATE` might be very similar.

I could have used OCaml code to generate the names consistently rather than using the C macros, at the time it seemed like a six of one, half-a-dozen of the other kind of situation.

**Comment by Timmmm** [comment]
Ah very helpful thanks!
